building with lotus API (#1190)
filecoin-project/filecoin-docs: content/en/build/building-with-lotus-APIs.md

---
title: Building with Lotus APIs

> description: Lotus node provides comprehensive features via its JSON-RPC APIs for developers and builders to build on the Filecoin network. Developers can use any API client library to connect to Lotus node to use the build-in Lotus APIs.

Reviewer: ```suggestion
description: Lotus offers a built-in JSON-RPC API for developers who need direct, protocol-level control for building tools and applications. Several client libraries are available.
```

menu:
    build:
        parent: "build-get-started"
weight: 40
---

> In the Filecoin ecosystem, there are various node solutions to connect to the Filecoin network and provide  functions via RPC API calls. Developers can either install and run their own Lotus node locally to interact with the Filecoin network, or connect to other Lotus nodes hosted by Glif or Infura.

Reviewer 2: ```suggestion
In the Filecoin ecosystem, there are various node solutions to connect to the Filecoin network and provide functions via RPC API calls. Developers can either install and run their own Lotus node locally to interact with the Filecoin network, or connect to other Lotus nodes hosted by Glif or Infura.
```

- [filecoin.js](https://filecoin-shipyard.github.io/filecoin.js/)
- [filecoin-js-signer](https://github.com/blitslabs/filecoin-js-signer)
- [Filecoin Signing Tools](https://github.com/Zondax/filecoin-signing-tools)

In this tutorial, we will run a local Lotus node and use the filecoin.js library to walk through the following examples: 

- Setting up a local Lotus node and connecting to [Calibration Testnet](https://docs.filecoin.io/networks/#calibration)
- Using filecoin.js to write code to interact with Filecoin Network to
   - query basic blockchain info
   - manage Lotus wallet
   - manage light wallet
   - make a storage deal

## Prerequisites

Before getting started, you should be familiar with:

- [How Filecoin works](https://docs.filecoin.io/about-filecoin/how-filecoin-works/)
- [How to install a Lotus node for the Calibration network](https://lotu.sh/lotus/install/prerequisites/#supported-platforms)
- [How to use Lotus command line](https://lotu.sh/lotus/manage/lotus-cli/)

> Also, make sure you have all the dependencies installed:
> 
> - Fully Synced Calibration TestNet Lotus node
> - node.js
> - npm

Reviewer 2: ```suggestion
Also, make sure you have all the dependencies installed:

- A fully-synced Lotus node on the calibration testnet
- Node.js
- NPM
```

## Setup

Before you start writing code to interact with hosted local Lotus node, you should make sure everything is set up properly. Especially ensure you have a local Lotus node fully synced for Calibration TestNet. 

### Set up lotus node

If you have not installed a Lotus node locally and fully synced for Calibration TestNet, you should go back and do it following the guild to [Install and launch a Lotus node](https://lotus.filecoin.io/docs/set-up/install/).

> 
> To sum up, you need to:

Reviewer 2: ```suggestion
```

> 1. Download Lotus source code for the latest release

Reviewer 2: ```suggestion
1. Download Lotus source code for the latest release:
```

   ```shell
   git clone https://github.com/filecoin-project/lotus.git
   cd lotus/
   git checkout <vX.X.X> # tag for a release, currently v1.15.0
   ```

> 2. Build and install Lotus for Calibnet

Reviewer 2: Calibnet should likely be in lowercase.

   ```shell

>    make clean calibnet # Calibration with min 32GiB sectors

Reviewer 2: This comment isn't necessary.

   sudo make install
   ```

> 3. Fast sync using latest snapshot (still need several hours to fully sync the chain)

Reviewer 2: Don't use brackets (). Move this part of the sentence somewhere else.

   ```shell
   lotus daemon --import-snapshot <path-to-calibration-snapshot>
   ```

> 4. [Create a wallet](https://lotu.sh/lotus/manage/manage-fil/#creating-a-wallet) and request some TestNet FIL from [Faucet](https://faucet.calibration.fildev.network/)

Reviewer 2: Why are you using `lotu.sh` links?

> 4. [Enable remote API access to your Lotus node](https://lotu.sh/developers/api-access/)

Reviewer 2: Again, don't use `lotu.sh` links. Use `docs.lotus.filecoin`

Author: I will update all the lotu.sh.
This PR was created before we started using `docs.lotus.filecoin`

If everything set up correctly, you should be able to verify to:

- make sure that everything is working correctly, check the status of your local node:

  ```shell with-output
  #Check if the node sync to the highest tipset
  lotus sync wait
  ```

  ```shell
  Worker: 10161; Base: 778820; Target: 778820 (diff: 0)
  State: complete; Current Epoch: 778820; Todo: 0
  
  Done!
  ```

- Check your wallet balance

  ```shell with-output
  #Check wallet balance
  lotus wallet list 
  ```

  ```shell
  Address                                    Balance        Nonce  
  t154xvuihhicgluafwmohwzwmtmqp44pwwgewyvma  100 FIL          0

>   ```

Reviewer 2: This section shouldn't be split into bullet points. Write something like:

```
To ensure everything has been set up correct, check the status of your node, and then check the balance of you wallet.
```

Then show them how to do both those things. Don't chop it up into bullet points.

> ### Create a node.js project

Reviewer 2: Node.js should be capitalized.

> Let's also set up a Node.js project with necessary dependencies, we will write basic javascript functions to focus more on learning the code to interact with Filecoin Network. 

Reviewer 2: ```suggestion
```

This sentence is redundant.

> 1. Create a new node.js project

Reviewer 2: ```suggestion
1. Create a new Node.js project:
```

Capitalize Node.js. End list items with colons.

   ```shell
   mkdir build-with-lotus && cd build-with-lotus

>    npm init #Initialize an project

Reviewer 2: ```suggestion
   npm init
```

Try not to comment your code. If something need explaining, explain it in a regular paragraph.

   ```
   
2. Install the fielcoin.js library

   ```shell
   npm install filecoin.js
   ```

> 

Reviewer 2: Make sure there's only one space at the end of sections. This was likely just a typo.

### Test a simple function call

Now you're ready to write code to connect to your Lotus node and interact with the Lotus APIs. Let's write a simple script to query the Lotus version to verify if everything works as expected.

> 1. Create a `chainDataQuery.mjs` file in your project and add the following code. We use the `mjs` file format because it supports import statements.

Reviewer 2: End items with a colon.

    ```javascript
    //1. Import necessary packages
    import { HttpJsonRpcConnector, LotusClient } from "filecoin.js";

    //2. Use the local node URL to create a connection to your Lotus node
    const localNodeUrl = "http://127.0.0.1:1234/rpc/v0";
    const localConnector = new HttpJsonRpcConnector({ url: localNodeUrl });

    //3. lotusClient exposes all Lotus APIs
    const lotusClient = new LotusClient(localConnector);
    const version = await lotusClient.common.version();
    console.log(version);
    ```

2. Run the following command in your project directory:

    ```shell with-output
    ~/build-with-lotus $ node basicQuery.mjs
    ```

    ```shell
    {
      Version: '1.15.0+calibnet+git.0ac1bbc7a',
      APIVersion: 66816,
      BlockDelay: 30
    }

>     ```

Reviewer 2: The `with-output` class isn't enabled on this site. Instead, write something like:

````
```shell
cat text.txt
```

This should output something like:

```shelll
Hello world!
```
````

Author: This will add a lot more unnecessary text in the tutorial. It is quite obvious that those blocks are output so that I will leave it as it is now.

> If everything is set up correctly, your code will be able to connect to the Lotus node and return the chain data as shown above. Now, you're ready to explore more API features.

Reviewer 2: What if stuff doesn't work? Where should the reader look for more information?

## Query chain data

Blockchain data is a fundamental part of any blockchain network. Using Lotus APIs, you can query chain data, including tipsets, blocks, messages, network status, storage deals, storage provider details, and more.

Let's look at some common chain data-related packages in the filecoin.js library's LotusClient. 

- `LotusClient.common` - methods to get node info
- `LotusClient.chain` - methods to interact with the blockchain
- `LotusClient.wallet` - methods to manage wallet on the Lotus node
- `LotusClient.client` - methods to interact with the storage and retrieval markets as a client
- `LotusClient.paych` - methods to interact with and manage payment channels
- `LotusClient.state` - methods to query, inspect, and interact with chain state

> - more ...

Reviewer 2: ```suggestion
```

This isn't important.

> In the previous step, we already created a `chainDataQuery.mjs` file to demonstrate the basic steps to connect to the Lotus node and initialize a LotusClient to query chain data. Now let's write more code in there to learn other query functions that do not require auth-token to interact with the Lotus node.

Reviewer 2: ```suggestion
In the previous step, we created a `chainDataQuery.mjs` file to demonstrate the basic steps to connect to the Lotus node and initialize a LotusClient to query chain data. Now, let's write more code in there to learn other query functions that do not require auth-token to interact with the Lotus node.
```

1. Get current chain head

   This function will return the current head of the chain, including current block height, a list of blocks, and a list of CIDs.
     ```javascript with-output
      //Query the current block head
      const chainHead = await lotusClient.chain.getHead();
      console.log(chainHead);
     ```

      ```shell
      {
        Cids: [
          {'/': 'bafy2bzacecdkonmhngylnnhrk4azkg2wkgcm6cnm5qn5sk4ww5cszjlvkgkd6'},
          {'/': 'bafy2bzaceaglcpzhd5gfrzdyt7ce3e5asnbfz3s3stbqyxniziny5snewbpbg'}
          ],
        Blocks: [
          {...},
          {...}
          ],
        Height: 781268
      }

>       ```

Reviewer 2: There's a lot of lines here for a single list item. Should this section be split into smaller headers maybe? Or just remove the `This function will return [...]` bit and tell the user what they need to do. If you need to explain what's going on in the code, do it after you've told them what the code should look like.

Author: If I fully post the chain head body here, it will be so many lines (might be 50+ for each block). So I just `...` as an ellipsis for block body. Do you have any suggestions that I can use instead of `...` or it is acceptable?

Author: Fixed it.  Thank you, Mosh

2. Get messages in a block

    ```javascript with-output
    //Query all the messages in a block
    const tipSet = await lotusClient.chain.getTipSetByHeight(781267);
    const messages = await lotusClient.chain.getBlockMessages(tipSet.Cids[0]);
    console.log(messages);
    ```

    ```shell
    {
      BlsMessages: [
        {
          Version: 0,
          To: 't3ukhj6tpkgxjknu54opiaej2vrjz7nh7gzodkqlhpfphc6gxkogzmojv2cnlpgcuwkvnyhloctnc6lmlvceuq',
          From: 't3ws4jhb7g4n5s7t3mwsc2enzfija2kzo3p7vkqdqbiaqrabqhmfxzzy3knrvn3ykp75okab6pse2nxud3ineq',
          ...
        }
      ],
      SecpkMessages: [ { Message: [Object], Signature: [Object], CID: [Object] } ],
      Cids: [
        {'/': 'bafy2bzaceapv6ms4m3x4gdaefmpilgsqfhfsrvrciw4iux4n7va47es4vudpa'},
        {'/': 'bafy2bzaceaxaz5ews6a2jnwmeafhi2p5xpwjz7kogilakblp3q3wpxa3ax2xy'}
      ]
    }
    ```

3. Get wallet balance

    With a given TestNet wallet address, you can query its balance by calling lotusClient.wallet.balance() function.

    ```javascript
    //Query Wallet balance
    const walletBalance = await lotusClient.wallet.balance("t1ne72cbn6r55wea7ifjv4ypyti7t2df5dumsjhzq");

>     ```

Reviewer 2: See previous comment about explaining your code _after_ you've shown it to the user.

Author: Fix it.

Author: Fix it.

1. Get current chain head:

   We can add the following code in `chainDataQuery.mjs` to query current head of the chain. 

     ```javascript with-output
      //Query the current block head
      const chainHead = await lotusClient.chain.getHead();
      console.log(chainHead);
     ```
   As a result, it will return the current head of the chain, including current block height, a list of blocks, and a list of CIDs.
      ```shell
      {
        Cids: [
          {'/': 'bafy2bzacecdkonmhngylnnhrk4azkg2wkgcm6cnm5qn5sk4ww5cszjlvkgkd6'},
          {'/': 'bafy2bzaceaglcpzhd5gfrzdyt7ce3e5asnbfz3s3stbqyxniziny5snewbpbg'}
          ],
        Blocks: [
          {...},
          # long list of additional blocks
          ],
        Height: 781268
      }
      ```

2. Get messages in a block

    ```javascript with-output
    //Query all the messages in a block
    const tipSet = await lotusClient.chain.getTipSetByHeight(781267);
    const messages = await lotusClient.chain.getBlockMessages(tipSet.Cids[0]);
    console.log(messages);

>     ```
>     Using the above code in`chainDataQuery.mjs` , we can also retrieve messages from a certain block height. And the returned message body will look like the following .
>     ```shell

Reviewer 2: ````suggestion

    ```
    Using the above code in`chainDataQuery.mjs`, we can also retrieve messages from a certain block height. And the returned message body will look like the following.

    ```shell
````

1. [Lotus wallets](#lotus-wallets) that are held directly in Lotus node.
2. [Light wallets](#light-wallet) that are detached from any node, and are managed by key-pairs.

### Lotus wallets

Lotus wallets are created and hosted in the Lotus node. When the Lotus node is running, the wallet is opened in the node and you can manage your wallet's addresses and balances using the command line or API calls. You will need an authorization token with sign permissions to interact with a wallet. Authorization tokens are often called _auth tokens_.

1. First, let's make sure your local Lotus node is running and the wallet address contains some testnet FIL.

    ```shell with-output
    lotus wallet list 
    ```

    This will output something like:

    ```shell
    Address                                    Balance        Nonce  
    t154xvuihhicgluafwmohwzwmtmqp44pwwgewyvma  100 FIL          0
    ```

1. Next, generate an authorization token with _sign_ permission:

    ```shell with-output
    lotus auth create-token --perm sign
    ```

    This will output your auth token. Save this somewhere, we'll be using it later:

    ```shell

>     eyJhbGciOiJIUzI1NiIsInRaaaaaa.xxxxxxx.bbbbbbbbbbbvq1W0ZjqrXHygd6fBRk

Reviewer 2: ```suggestion
eyJhbGciOiJIUzI1NiIsInRaaaaaa.xxxxxxx.bbbbbbbbbbbvq1W0ZjqrXHygd6fBRk
```
